File: 04_Binary_Search/02_Search_a_2D_Matrix.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        m , n = len(matrix), len(matrix[0])
        i, j = 0, n-1

        while i < m and j >= 0:
            x = matrix[i][j]
            if x == target:
                return True
            elif target > x:
                i += 1
            elif target < x:
                j -= 1
            else:
                return False
        
        return False

File: 04_Binary_Search/02_Search_a_2D_Matrix.cpp
class Solution
{
public:
    bool searchMatrix(vector<vector<int>> &matrix, int target)
    {
        int m = matrix.size(), n = matrix[0].size();
        int i = 0, j = n - 1;

        while (i < m && j >= 0)
        {
            int x = matrix[i][j];
            if (x == target)
                return true;
            else if (x < target)
                i++;
            else if (x > target)
                j--;
            else
                return false;
        }

        return false;
    }
};